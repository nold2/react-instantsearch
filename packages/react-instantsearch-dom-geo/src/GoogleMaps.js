import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { LatLngPropType, BoundingBoxPropType } from './propTypes';

export const GOOGLE_MAPS_CONTEXT = '__ais_geo_search__google_maps__';

class GoogleMaps extends Component {
  static propTypes = {
    cx: PropTypes.func.isRequired,
    google: PropTypes.object.isRequired,
    initialZoom: PropTypes.number.isRequired,
    initialPosition: LatLngPropType.isRequired,
    mapOptions: PropTypes.object.isRequired,
    refine: PropTypes.func.isRequired,
    position: LatLngPropType,
    boundingBox: BoundingBoxPropType,
    children: PropTypes.node,
  };

  static childContextTypes = {
    [GOOGLE_MAPS_CONTEXT]: PropTypes.shape({
      google: PropTypes.object,
      instance: PropTypes.object,
    }),
  };

  isUserInteraction = true;
  isPendingRefine = false;
  listeners = [];

  state = {
    isMapReady: false,
  };

  getChildContext() {
    const { google } = this.props;

    return {
      [GOOGLE_MAPS_CONTEXT]: {
        instance: this.instance,
        google,
      },
    };
  }

  componentDidMount() {
    const { google, mapOptions } = this.props;

    this.instance = new google.maps.Map(this.element, {
      mapTypeControl: false,
      fullscreenControl: false,
      streetViewControl: false,
      clickableIcons: false,
      zoomControlOptions: {
        position: google.maps.ControlPosition.LEFT_TOP,
      },
      ...mapOptions,
    });

    google.maps.event.addListenerOnce(
      this.instance,
      'idle',
      this.setupListenersWhenMapIsReady
    );
  }

  componentDidUpdate() {
    const {
      google,
      initialZoom,
      initialPosition,
      boundingBox,
      position,
    } = this.props;

    if (boundingBox) {
      this.lockUserInteration(() => {
        this.instance.fitBounds(
          new google.maps.LatLngBounds(
            boundingBox.southWest,
            boundingBox.northEast
          ),
          0
        );
      });
    }

    if (!boundingBox) {
      const initialMapPosition = position || initialPosition;

      this.lockUserInteration(() => {
        this.instance.setZoom(initialZoom);
        this.instance.setCenter(initialMapPosition);
      });
    }
  }

  componentWillUnmount() {
    this.listeners.forEach(listener => {
      listener.remove();
    });

    this.listeners = [];
  }

  setupListenersWhenMapIsReady = () => {
    const { refine } = this.props;

    this.setState(() => ({
      isMapReady: true,
    }));

    const onChange = () => {
      if (this.isUserInteraction) {
        this.isPendingRefine = true;
      }
    };

    this.listeners.push(this.instance.addListener('center_changed', onChange));
    this.listeners.push(this.instance.addListener('zoom_changed', onChange));
    this.listeners.push(this.instance.addListener('dragstart', onChange));

    this.listeners.push(
      this.instance.addListener('idle', () => {
        if (this.isUserInteraction && this.isPendingRefine) {
          this.isPendingRefine = false;

          const bounds = this.instance.getBounds();

          refine({
            northEast: bounds.getNorthEast().toJSON(),
            southWest: bounds.getSouthWest().toJSON(),
          });
        }
      })
    );
  };

  lockUserInteration(functionThatAlterTheMapPosition) {
    this.isUserInteraction = false;
    functionThatAlterTheMapPosition();
    this.isUserInteraction = true;
  }

  render() {
    const { cx, children } = this.props;
    const { isMapReady } = this.state;

    return (
      <div className={cx('')}>
        <div ref={c => (this.element = c)} className={cx('map')} />
        {isMapReady && children}
      </div>
    );
  }
}

export default GoogleMaps;
