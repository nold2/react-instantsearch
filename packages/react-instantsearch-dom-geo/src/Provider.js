import { Component } from 'react';
import PropTypes from 'prop-types';
// @TODO: Update this import when the package is correctly split:
// import { connectGeoSearch } from 'react-instantsearch-dom';
import { connectGeoSearch } from '../../react-instantsearch/connectors';
import { LatLngPropType, BoundingBoxPropType } from './propTypes';

export const STATE_CONTEXT = '__ais_geo_search__state__';

export class Provider extends Component {
  static propTypes = {
    hits: PropTypes.arrayOf(PropTypes.object).isRequired,
    isRefinedWithMap: PropTypes.bool.isRequired,
    refine: PropTypes.func.isRequired,
    children: PropTypes.func.isRequired,
    position: LatLngPropType,
    currentRefinement: BoundingBoxPropType,
  };

  static childContextTypes = {
    [STATE_CONTEXT]: PropTypes.shape({
      hasMapMoveSinceLastRefine: PropTypes.bool.isRequired,
      setMapMoveSinceLastRefine: PropTypes.func.isRequired,
    }).isRequired,
  };

  state = {
    hasMapMoveSinceLastRefine: false,
  };

  getChildContext() {
    const { hasMapMoveSinceLastRefine } = this.state;

    return {
      [STATE_CONTEXT]: {
        setMapMoveSinceLastRefine: this.setMapMoveSinceLastRefine,
        hasMapMoveSinceLastRefine,
      },
    };
  }

  setMapMoveSinceLastRefine = next => {
    const { hasMapMoveSinceLastRefine } = this.state;

    if (hasMapMoveSinceLastRefine === next) {
      return;
    }

    this.setState(() => ({
      hasMapMoveSinceLastRefine: next,
    }));
  };

  render() {
    const {
      hits,
      isRefinedWithMap,
      position,
      currentRefinement,
      refine,
      children,
    } = this.props;

    const { hasMapMoveSinceLastRefine } = this.state;

    return children({
      setMapMoveSinceLastRefine: this.setMapMoveSinceLastRefine,
      hits,
      isRefinedWithMap,
      hasMapMoveSinceLastRefine,
      position,
      currentRefinement,
      refine,
    });
  }
}

export default connectGeoSearch(Provider);
