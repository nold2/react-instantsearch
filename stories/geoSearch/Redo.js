import React from 'react';
import PropTypes from 'prop-types';
import Button from './Button';

const Redo = (_, { hasMapMoveSinceLastRefine, refineWithBoudingBox }) => (
  <Button
    className="redo"
    onClick={() => refineWithBoudingBox()}
    disabled={!hasMapMoveSinceLastRefine}
  >
    Redo search here
  </Button>
);

Redo.contextTypes = {
  refineWithBoudingBox: PropTypes.func.isRequired,
  hasMapMoveSinceLastRefine: PropTypes.bool.isRequired,
};

export default Redo;
