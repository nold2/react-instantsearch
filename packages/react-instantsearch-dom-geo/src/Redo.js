import React, { Component } from 'react';
import PropTypes from 'prop-types';
// @TODO: Update this import when the package is correctly split:
// import { createClassNames } from 'react-instantsearch-dom';
import createClassNames from '../../react-instantsearch/src/components/createClassNames';
// @TODO: Update this import when the package is correctly split:
// import { translatable } from 'react-instantsearch-core';
import translatable from '../../react-instantsearch/src/core/translatable';
import { STATE_CONTEXT } from './Provider';
import { GOOGLE_MAPS_CONTEXT } from './GoogleMaps';

const cx = createClassNames('GeoSearch');

export class Redo extends Component {
  static propTypes = {
    translate: PropTypes.func.isRequired,
  };

  static contextTypes = {
    [STATE_CONTEXT]: PropTypes.shape({
      isRefineOnMapMove: PropTypes.bool.isRequired,
      toggleRefineOnMapMove: PropTypes.func.isRequired,
      hasMapMoveSinceLastRefine: PropTypes.bool.isRequired,
    }).isRequired,
    [GOOGLE_MAPS_CONTEXT]: PropTypes.shape({
      refineWithBoundingBox: PropTypes.func.isRequired,
    }).isRequired,
  };

  getStateContext() {
    return this.context[STATE_CONTEXT];
  }

  getGoogleMapsContext() {
    return this.context[GOOGLE_MAPS_CONTEXT];
  }

  componentDidMount() {
    const { isRefineOnMapMove, toggleRefineOnMapMove } = this.getStateContext();

    if (isRefineOnMapMove) {
      toggleRefineOnMapMove();
    }
  }

  render() {
    const { translate } = this.props;
    const { hasMapMoveSinceLastRefine } = this.getStateContext();
    const { refineWithBoundingBox } = this.getGoogleMapsContext();

    return (
      <div className={cx('control')}>
        <button
          className={cx('redo', !hasMapMoveSinceLastRefine && 'redo--disabled')}
          disabled={!hasMapMoveSinceLastRefine}
          onClick={refineWithBoundingBox}
        >
          {translate('redo')}
        </button>
      </div>
    );
  }
}

export default translatable({
  redo: 'Redo search here',
})(Redo);
