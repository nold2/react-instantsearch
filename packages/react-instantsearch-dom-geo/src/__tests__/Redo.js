import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { STATE_CONTEXT } from '../Provider';
import { GOOGLE_MAPS_CONTEXT } from '../GoogleMaps';
import { Redo } from '../Redo';

Enzyme.configure({ adapter: new Adapter() });

describe('Redo', () => {
  const defaultProps = {
    translate: x => x,
  };

  const defaultContext = {
    [STATE_CONTEXT]: {
      isRefineOnMapMove: true,
      hasMapMoveSinceLastRefine: false,
      toggleRefineOnMapMove: () => {},
    },
    [GOOGLE_MAPS_CONTEXT]: {
      refineWithBoundingBox: () => {},
    },
  };

  const getStateContext = context => context[STATE_CONTEXT];
  const getGoogleMapsContext = context => context[GOOGLE_MAPS_CONTEXT];

  it('expect to render correctly', () => {
    const props = {
      ...defaultProps,
    };

    const context = {
      ...defaultContext,
    };

    const wrapper = shallow(<Redo {...props} />, {
      context,
    });

    expect(wrapper.find('button').prop('disabled')).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('expect to render correctly when map has moved', () => {
    const props = {
      ...defaultProps,
    };

    const context = {
      ...defaultContext,
      [STATE_CONTEXT]: {
        ...getStateContext(defaultContext),
        hasMapMoveSinceLastRefine: true,
      },
    };

    const wrapper = shallow(<Redo {...props} />, {
      context,
    });

    expect(wrapper.find('button').prop('disabled')).toBe(false);
    expect(wrapper).toMatchSnapshot();
  });

  it('expect to disable refine on map move onDidMount', () => {
    const props = {
      ...defaultProps,
    };

    const context = {
      ...defaultContext,
      [STATE_CONTEXT]: {
        ...getStateContext(defaultContext),
        toggleRefineOnMapMove: jest.fn(),
      },
    };

    const wrapper = shallow(<Redo {...props} />, {
      disableLifecycleMethods: true,
      context,
    });

    expect(
      getStateContext(context).toggleRefineOnMapMove
    ).toHaveBeenCalledTimes(0);

    wrapper.instance().componentDidMount();

    expect(
      getStateContext(context).toggleRefineOnMapMove
    ).toHaveBeenCalledTimes(1);
  });

  it('expect to only disable refine on map move when value is true onDidMount', () => {
    const props = {
      ...defaultProps,
    };

    const context = {
      ...defaultContext,
      [STATE_CONTEXT]: {
        ...getStateContext(defaultContext),
        isRefineOnMapMove: false,
        toggleRefineOnMapMove: jest.fn(),
      },
    };

    const wrapper = shallow(<Redo {...props} />, {
      disableLifecycleMethods: true,
      context,
    });

    expect(
      getStateContext(context).toggleRefineOnMapMove
    ).toHaveBeenCalledTimes(0);

    wrapper.instance().componentDidMount();

    expect(
      getStateContext(context).toggleRefineOnMapMove
    ).toHaveBeenCalledTimes(0);
  });

  it('expect to call refineWithBoundingBox on button click', () => {
    const props = {
      ...defaultProps,
    };

    const context = {
      ...defaultContext,
      [GOOGLE_MAPS_CONTEXT]: {
        ...getGoogleMapsContext(defaultContext),
        refineWithBoundingBox: jest.fn(),
      },
    };

    const wrapper = shallow(<Redo {...props} />, {
      context,
    });

    wrapper.find('button').simulate('click');

    expect(
      getGoogleMapsContext(context).refineWithBoundingBox
    ).toHaveBeenCalledTimes(1);
  });
});
