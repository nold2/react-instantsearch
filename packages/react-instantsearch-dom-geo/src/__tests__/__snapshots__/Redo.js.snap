// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redo expect to render correctly 1`] = `
<div
  className="ais-GeoSearch-control"
>
  <button
    className="ais-GeoSearch-redo ais-GeoSearch-redo--disabled"
    disabled={true}
    onClick={[Function]}
  >
    redo
  </button>
</div>
`;

exports[`Redo expect to render correctly when map has moved 1`] = `
<div
  className="ais-GeoSearch-control"
>
  <button
    className="ais-GeoSearch-redo"
    disabled={false}
    onClick={[Function]}
  >
    redo
  </button>
</div>
`;
